flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java

```diff
@@ -0,0 +1,167 @@
+ *     <li>{@link ShuffleEnvironment#releasePartitions(Collection)} is called outside of the producer thread,
+ *     e.g. to manage the lifecycle of BLOCKING result partitions which can outlive their producers.</li>
+ * </ol>
+ * The partitions, which currently still occupy local resources, can be queried with
+ * {@link ShuffleEnvironment#getPartitionsOccupyingLocalResources}.
+ *
+ * <h2>Input gate management.</h2>
+ *
+ * <p>The interface implements a factory for the input gates: {@link ShuffleEnvironment#createInputGates}.
+ * The created gates are grouped per owner. The owner is responsible for the gates' lifecycle from the moment of creation.
+ *
+ * <p>When the input gates are created, it can happen that not all consumed partitions are known at that moment
+ * e.g. because their producers have not been started yet. Therefore, the {@link ShuffleEnvironment} provides
+ * a method {@link ShuffleEnvironment#updatePartitionInfo} to update them externally, when the producer becomes known.
+ * The update mechanism has to be threadsafe because the updated gate can be read concurrently from a different thread.
+ *
+ * @param <P> type of provided result partition writers
+ * @param <G> type of provided input gates
+ */
+public interface ShuffleEnvironment<P extends ResultPartitionWriter, G extends InputGate> extends AutoCloseable {
+
+	/**
+	 * Start the internal related services before using the shuffle service environment.
+	 *
+	 * @return a port to connect for the shuffle data exchange, -1 if only local connection is possible.
+	 */
+	int start() throws IOException;
+
+	/**
+	 * Create a context of the shuffle input/output owner used to create partitions or gates belonging to the owner.
+	 *
+	 * <p>This method has to be called only once to avoid duplicated internal metric group registration.
+	 *
+	 * @param ownerName the owner name, used for logs
+	 * @param executionAttemptID execution attempt id of the producer or consumer
+	 * @param parentGroup parent of shuffle specific metric group
+	 * @return context of the shuffle input/output owner used to create partitions or gates belonging to the owner
+	 */
+	ShuffleIOOwnerContext createShuffleIOOwnerContext(
+		String ownerName,
+		ExecutionAttemptID executionAttemptID,
+		MetricGroup parentGroup);
+
+	/**
+	 * Factory method for the {@link ResultPartitionWriter ResultPartitionWriters} to produce result partitions.
+	 *
+	 * <p>The order of the {@link ResultPartitionWriter ResultPartitionWriters} in the returned collection
+	 * should be the same as the iteration order of the passed {@code resultPartitionDeploymentDescriptors}.
+	 *
+	 * @param ownerContext the owner context relevant for partition creation
+	 * @param resultPartitionDeploymentDescriptors descriptors of the partition, produced by the owner
+	 * @return collection of the {@link ResultPartitionWriter ResultPartitionWriters}
+	 */
+	Collection<P> createResultPartitionWriters(
```

Comment:
have you considered moving this method and `createInputGates` into the `ShuffleIOOwnerContext`, and only pass this into the task instead of the entire `ShuffleEnvironment`? The task only requires access to these 2 methods, while any other users of the `ShuffleEnvironment` doesn't need these.

Reply:
This is an interesting idea, we can consider it as a follow-up issue.